be/src/olap/rowset/segment_v2/binary_prefix_page.cpp

```diff
@@ -0,0 +1,250 @@
+        _cur_pos = _num_values;
+        return Status::OK();
+    }
+
+    size_t restart_point_index = pos / RESTART_POINT_INTERVAL;
+    RETURN_IF_ERROR(_seek_to_restart_point(restart_point_index));
+    while (_cur_pos < pos) {
+        _cur_pos++;
+        RETURN_IF_ERROR(_read_next_value());
+    }
+    return Status::OK();
+}
+
+Status BinaryPrefixPageDecoder::seek_at_or_after_value(const void* value, bool* exact_match) {
+    DCHECK(_parsed);
+    Slice target = *reinterpret_cast<const Slice*>(value);
+
+    uint32_t left = 0;
+    uint32_t right = _num_restarts;
+    // find the first restart point >= target. after loop,
+    // - left == index of first restart point >= target when found
+    // - left == _num_restarts when not found (all restart points < target)
+    while (left < right) {
+        uint32_t mid = left + (right - left) / 2;
+        // read first entry at restart point `mid`
+        RETURN_IF_ERROR(_seek_to_restart_point(mid));
+        Slice mid_entry(_current_value);
+        if (mid_entry.compare(target) < 0) {
+            left = mid + 1;
+        } else {
+            right = mid;
+        }
+    }
+
+    // then linear search from the last restart pointer < target.
+    // when left == 0, all restart points >= target, so search from first one.
+    // otherwise search from the last restart point < target, which is left - 1
+    uint32_t search_index = left > 0 ? left - 1 : 0;
+    RETURN_IF_ERROR(_seek_to_restart_point(search_index));
+    while (true) {
+        int cmp = Slice(_current_value).compare(target);
+        if (cmp >= 0) {
+            *exact_match = cmp == 0;
+            return Status::OK();
+        }
+        _cur_pos++;
+        RETURN_IF_ERROR(_read_next_value());
+    }
+}
+
+Status BinaryPrefixPageDecoder::_copy_current_to_output(MemPool* mem_pool, Slice* output) {
+    output->size = _current_value.size();
+    if (output->size > 0) {
+        output->data = (char*) mem_pool->allocate(output->size);
+        if (output->data == nullptr) {
+            return Status::MemoryAllocFailed(
+                Substitute("failed to allocate $0 bytes", output->size));
+        }
+        memcpy(output->data, _current_value.data(), output->size);
+    }
+    return Status::OK();
+}
+
+Status BinaryPrefixPageDecoder::next_batch(size_t* n, ColumnBlockView* dst) {
+    DCHECK(_parsed);
+    if (PREDICT_FALSE(*n == 0 || _cur_pos >= _num_values)) {
+        *n = 0;
+        return Status::OK();
+    }
+    size_t i = 0;
+    size_t max_fetch = std::min(*n, static_cast<size_t>(_num_values - _cur_pos));
+    auto out = reinterpret_cast<Slice*>(dst->data());
+
+    // first copy the current value to output
+    RETURN_IF_ERROR(_copy_current_to_output(dst->pool(), out));
+    i++;
+    out++;
+
+    // read and copy remaining values
+    for (; i < max_fetch; ++i) {
+        _cur_pos++;
+        RETURN_IF_ERROR(_read_next_value());
+        RETURN_IF_ERROR(_copy_current_to_output(dst->pool(), out));
```

Comment:
Is it possible and necessary to merge these two memory copies to one copy?

Reply:
Yes. I has merged two memory copies to one copy.